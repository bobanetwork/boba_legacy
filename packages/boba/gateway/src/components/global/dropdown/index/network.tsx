import React, { useState, useCallback } from 'react'
import ArrowDown from '../../../../images/icons/arrowdown.svg'
import {
  DropdownContainer,
  Header,
  IconContainer,
  Option,
  DefaultIcon,
  DropdownBody,
  Icon,
  DropdownContent,
} from '../themes/networkStyles'
interface IDropdownItem {
  value: string
  label: string
  imgSrc?: string
}

export interface IDropdownProps {
  items: IDropdownItem[]
  defaultItem: IDropdownItem
  onItemSelected?: (item: IDropdownItem) => void
}

export const Dropdown: React.FC<IDropdownProps> = ({
  items,
  defaultItem,
  onItemSelected,
}) => {
  const [selectedItem, setSelectedItem] = useState<IDropdownItem>(defaultItem)
  const [isOpen, setIsOpen] = useState<boolean>(false)

  const handleDropdown = useCallback(() => {
    setIsOpen(!isOpen)
  }, [isOpen])

  const selectItem = useCallback(
    (item: IDropdownItem) => {
      setSelectedItem(item)
      setIsOpen(false)
      onItemSelected && onItemSelected(item)
    },
    [onItemSelected]
  )

  return (
    <DropdownContainer className="dropdown">
      <Header onClick={handleDropdown}>
        <Option>
          {selectedItem.imgSrc && (
            <IconContainer>
              {selectedItem.imgSrc !== 'default' && (
                <Icon src={selectedItem.imgSrc} alt={selectedItem.label} />
              )}
              {selectedItem.imgSrc === 'default' && <DefaultIcon />}
            </IconContainer>
          )}
          {selectedItem.label}

          {isOpen ? <img src={ArrowDown} /> : <img src={ArrowDown} />}
        </Option>
      </Header>
      {isOpen && (
        <DropdownBody>
          <DropdownContent>
            {items.map((item, index) => (
              <Option
                key={index}
                onClick={() => selectItem(item)}
                style={{
                  flexDirection: 'row',
                  display: 'flex',
                  alignItems: 'center',
                  justifyContent: 'flex-start',
                }}
              >
                {item.imgSrc && (
                  <IconContainer style={{ alignContent: 'center' }}>
                    {item.imgSrc !== 'default' && (
                      <Icon
                        style={{ alignContent: 'center' }}
                        src={item.imgSrc}
                        alt={item.label}
                      />
                    )}
                    {item.imgSrc === 'default' && <DefaultIcon />}
                  </IconContainer>
                )}
                {item.label}
              </Option>
            ))}
          </DropdownContent>
        </DropdownBody>
      )}
    </DropdownContainer>
  )
}
