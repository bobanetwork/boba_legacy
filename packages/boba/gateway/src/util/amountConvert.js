/*
Copyright 2019-present OmiseGO Pte Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */

// we use BigNumber here for decimal support
import BigNumber from 'bignumber.js';

export function logAmount (amount, power, truncate = 0) {

  const x = new BigNumber(amount);
  const exp = new BigNumber(10).pow(power);

  const calculated = x.div(exp);

  if(truncate > 0)
  	return calculated.toFixed(truncate);
  else
  	return calculated.toFixed();
}

/*Takes a value such as 3.92 and converts it into
a BigNumber in wei

Duplicates

ethers.utils.parseUnits( valueString , decimalsOrUnitName )   =>   BigNumber
*/

export function powAmount (amount, decimals) {

  const x = new BigNumber(amount)
  const exp = new BigNumber(10).pow(decimals)

  const calculated = x.multipliedBy(exp)
  return calculated.toFixed(0)
}

/* more clearly named version of this */
export function toWei_String(amount, decimals) {

  const x = new BigNumber(amount)
  const exp = new BigNumber(10).pow(decimals)

  const calculated = x.multipliedBy(exp)
  return calculated.toFixed(0)
}

export function amountToUsd(amount, lookupPrice, token) {
  if (token.symbol === 'ETH' && !!lookupPrice['ethereum']) {
    return amount * lookupPrice['ethereum'].usd
  } else if (token.symbol === 'BOBA' && !!lookupPrice[ 'boba-network' ]) {
    return amount * lookupPrice['boba-network'].usd
  } else if (token.symbol === 'OLO' && !!lookupPrice[ 'oolongswap' ]) {
    return amount * lookupPrice['oolongswap'].usd
  } else if (token.symbol === 'OMG' && !!lookupPrice[ 'omisego' ]) {
    return amount * lookupPrice['omisego'].usd
  } else if (token.symbol === 'USDC' && !!lookupPrice[ 'usd-coin' ]) {
    return amount * lookupPrice['usd-coin'].usd 
  } else if (!!lookupPrice[ token.symbol.toLowerCase() ]) {
    return amount * lookupPrice[token.symbol.toLowerCase()].usd
  } else {
    return 0
  }
}
