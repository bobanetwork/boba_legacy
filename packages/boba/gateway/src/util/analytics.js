import ReactGA from "react-ga4"
import { GA4_MEASUREMENT_ID } from "./constant";

export const init = () => {
  if (window.location.hostname === 'localhost') {
    return null;
  }
  if (GA4_MEASUREMENT_ID) {
    // init if the GA4 Measurement Id is available.
    ReactGA.initialize(GA4_MEASUREMENT_ID);
  }
}

export const sendPageView = (path) => {
  ReactGA.send({
    hitType: "pageview",
    page: path,
    title: path
  });
}


export default {
  init,
  sendPageView
}
