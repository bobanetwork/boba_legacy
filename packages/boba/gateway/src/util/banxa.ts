import { BANXA_URL } from './constant'

interface Props {
  symbol: string
  address: string
}

const FIAT_TYPE: string = 'USD'
const BLOCK_CHAIN: string = 'Boba Network'

export const prepareBanxaUrl = ({ symbol, address }: Props) => {
  const endpoint = BANXA_URL

  return `${endpoint}coinType=${symbol}&fiatType=${FIAT_TYPE}&blockchain=${BLOCK_CHAIN}&walletAddress=${address}`
}
