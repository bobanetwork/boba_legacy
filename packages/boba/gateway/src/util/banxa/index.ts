import axios from 'axios'

const BANXA_ENDPOINT = 'http://localhost:4000'

type FiatType = {
  fiat_code: string
  fiat_name: string
  fiat_symbol: string
}

type blockchainType = {
  code: string
  description: string
  is_default: boolean
}
type CoinType = {
  coin_code: string
  coin_name: string
  blockchains: blockchainType[]
}

type feesType = {
  name: string
  amount: number
  type: string
}
type TransactionFeesType = {
  fiat_code: string
  coin_code: string
  fees: feesType[]
}

type TransactionLimitType = {
  fiat_code: string
  min: number
  max: number
}
export type PaymentMethod = {
  id: number
  name: string
  description: string
  logo_url: string
  status: string
  type: string
  paymentType: string
  supported_agents: null
  supported_fiat: string[]
  supported_coin: string[]
  transaction_fees: TransactionFeesType[]
  transaction_limits: TransactionLimitType[]
}

export interface OrderParams {
  account_reference: string
  source: string
  source_amount: string | undefined
  target: string
  wallet_address: string
  return_url_on_success?: string
  payment_method_id: string
}

export const getStatics = async (): Promise<any> => {
  try {
    const response = await axios.get(`${BANXA_ENDPOINT}/statics`)
    return response.data
  } catch (error) {
    console.error(error)
  }
}

export const createOrder = (params: OrderParams) => {
  const parameters = {
    return_url_on_success: 'https://gateway.boba.network/?orderStatus=created',
    ...params,
  }

  axios
    .post(`${BANXA_ENDPOINT}/create-order`, parameters)
    .then((response) => {
      console.log(response.data)
    })
    .catch((error) => {
      console.error(error)
    })
}
