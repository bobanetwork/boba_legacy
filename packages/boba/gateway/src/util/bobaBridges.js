export const bobaBridges = [
  {
    "name": "Synapse",
    "type": "SYNAPSE",
    "link": "https://synapseprotocol.com/",
    "tokens": [
      "ETH",
      "nETH",
      "gOHM",
      "DAI",
      "USDC",
      "USDT",
      "SYN",
      "nUSD"
    ]
  },
  {
    "name": "Anyswap",
    "type": "ANYSWAP",
    "link": "https://anyswap.exchange/#/router",
    "tokens": [
      "MIM",
      "AVAX",
      "FRAX",
      "FTM",
      "FXS",
      "MATIC"
    ]
  },
  {
    "name": "Across",
    "type": "ACROSS",
    "link": "https://across.to/",
    "tokens": [
      "ETH",
      "BOBA",
      "USDC",
      "WBTC",
      "UMA"
    ]
  },
  {
    "name": "Celer",
    "type": "CELER",
    "link": "https://cbridge.celer.network/#/transfer",
    "tokens": [
      "ETH",
      "BOBA",
      "FRAX",
      "OLO"
    ]
  },
  {
    "name": "BoringDAO",
    "type": "BORINGDAO",
    "link": "https://oportal.boringdao.com/twoway",
    "tokens": [
      "USDT"
    ]
  },
  {
    "name": "PolyBridge",
    "type": "POLYBRIDGE",
    "link": "https://bridge.poly.network/",
    "tokens": [
      "BOBA"
    ]
  },
  {
    "name": "Symbiosis",
    "type": "SYMBIOSIS",
    "link": "https://app.symbiosis.finance/swap",
    "tokens": [
      "USDC"
    ]
  }
]
