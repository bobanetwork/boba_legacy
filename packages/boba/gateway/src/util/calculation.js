/*
  Varna - A Privacy-Preserving Marketplace
  Varna uses Fully Homomorphic Encryption to make markets fair. 
  Copyright (C) 2021 Enya Inc. Palo Alto, CA

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program. If not, see <https://www.gnu.org/licenses/>.
*/


const getCorrectResult = (type, num1, num2, result) => {
    const operations = {
        add: (num1, num2) => num1 + num2,
        sub: (num1, num2) => num1 - num2,
        div: (num1, num2) => num1 / num2,
        mul: (num1, num2) => num1 * num2
    };
    
    const temp_result = operations[type] ? operations[type](num1, num2) : result;
    return Math.abs(result - temp_result) > 1 ? temp_result : result;
};

const countDecimals = (num) => {
    const scientificRegex = /^-?\d+(?:\.\d+)?(?:[eE][+-]?\d+)?$/;
    const decimalRegex = /^-?\d+\.\d+$/;

    if (isNaN(num) || num === '') return 0; // Check if it's a valid number

    const str = num.toString(); // Convert number to String;


    // We check if number are cientific notation
    const scientificMatch = str.match(scientificRegex);
    if (scientificMatch) {
        const [, decimalPart, exponent] = scientificMatch[0].match(/^(-?\d+(?:\.\d+)?)(?:[eE]([+-]?\d+))?$/);
        return Math.max((decimalPart.split('.')[1] || '').length - (exponent || 0), 0);
    }

    // We check if number are decimal 
    const decimalMatch = str.match(decimalRegex);
    if (decimalMatch) {
        return (decimalMatch[0].split('.')[1] || '').length;
    }

    return 0;
}

const convertToInt = (num) => {
    const isExponential = num.toString().toUpperCase().split('E').length === 2;
    const decimalPlaces = isExponential ? countDecimals(num) : 0;
    const numWithoutDecimals = isExponential ? Math.round(num * Math.pow(10, decimalPlaces)) : num;
    return Number(numWithoutDecimals.toString().replace(".", ""));
}


var accAdd = function(num1, num2) {
    num1 = Number(num1);
    num2 = Number(num2);
    var dec1, dec2, times;
    try { dec1 = countDecimals(num1)+1; } catch (e) { dec1 = 0; }
    try { dec2 = countDecimals(num2)+1; } catch (e) { dec2 = 0; }
    times = Math.pow(10, Math.max(dec1, dec2));
    var result = (accMul(num1, times) + accMul(num2, times)) / times;
    return getCorrectResult("add", num1, num2, result);
};

var accSub = function(num1, num2) {
    num1 = Number(num1);
    num2 = Number(num2);
    var dec1, dec2, times;
    try { dec1 = countDecimals(num1)+1; } catch (e) { dec1 = 0; }
    try { dec2 = countDecimals(num2)+1; } catch (e) { dec2 = 0; }
    times = Math.pow(10, Math.max(dec1, dec2));
    var result = Number((accMul(num1, times) - accMul(num2, times)) / times);
    return getCorrectResult("sub", num1, num2, result);
};

var accDiv = function(num1, num2) {
    num1 = Number(num1);
    num2 = Number(num2);
    var t1 = 0, t2 = 0, dec1, dec2;
    try { t1 = countDecimals(num1); } catch (e) { }
    try { t2 = countDecimals(num2); } catch (e) { }
    dec1 = convertToInt(num1);
    dec2 = convertToInt(num2);
    var result = accMul((dec1 / dec2), Math.pow(10, t2 - t1));
    return getCorrectResult("div", num1, num2, result);
};

var accMul = function(num1, num2) {
    num1 = Number(num1);
    num2 = Number(num2);
    var times = 0, s1 = num1.toString(), s2 = num2.toString();
    try { times += countDecimals(s1); } catch (e) { }
    try { times += countDecimals(s2); } catch (e) { }
    var result = convertToInt(s1) * convertToInt(s2) / Math.pow(10, times);
    return getCorrectResult("mul", num1, num2, result);
};

export {
  accAdd,
  accSub,
  accDiv,
  accMul,
  countDecimals,
  convertToInt,
  getCorrectResult,
}