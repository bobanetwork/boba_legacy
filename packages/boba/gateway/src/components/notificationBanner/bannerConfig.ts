import { NETWORK } from 'util/network/network.util'

interface BannerContent {
  message?: string
  content?: string
}

/**
 * To show alert specific to network update the config.
 *
 * eg.
 *
   BannerConfig = {
      [NETWORK.ETHEREUM]: {
        message: `Alert text for etheruem network`,
        content: `Descriptive alert text for etheruem network`,
      }
   }
 *
 *
 */

export const BannerConfig: Record<string, BannerContent> = {
  [NETWORK.AVAX]: {
    message: `BobaAvax (Fuji) is being wound down & will no longer be available, starting October 31st`,
    content: `BobaAvax (Fuji) is being wound down & will no longer be available starting October 31st. For users of BobaAvax or BobaAvax applications you will need to transfer all your funds to Avalanche mainnet before October 31st or risk permanently losing access to any assets on BobaAvax.`,
  },
}
