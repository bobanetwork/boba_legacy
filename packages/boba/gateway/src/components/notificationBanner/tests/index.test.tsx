import React from 'react'
import NotificationBanner from 'components/notificationBanner'
import { render, screen, fireEvent } from '@testing-library/react'
import CustomThemeProvider from 'themes'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import { NETWORK, NETWORK_TYPE } from 'util/network/network.util'
import { BannerConfig } from '../bannerConfig'

jest.mock('../bannerConfig', () => ({
  BannerConfig: {
    ETHEREUM: {
      message: `Alert message goes here`,
      content: `Alert descriptive message goes here`,
    },
  },
}))

const data = BannerConfig[NETWORK.ETHEREUM]

const mockStore = configureStore()

const renderBanner = (props: any) => {
  return render(
    <Provider
      store={mockStore({
        ui: {
          theme: 'dark',
        },
        network: {
          activeNetwork: NETWORK.ETHEREUM,
          activeNetworkType: NETWORK_TYPE.MAINNET,
        },
      })}
    >
      <CustomThemeProvider>
        <NotificationBanner {...props} />
      </CustomThemeProvider>
    </Provider>
  )
}

describe('NotificationBanner ', () => {
  test('should match snapshot without message', () => {
    const { asFragment } = renderBanner({})
    expect(asFragment()).toMatchSnapshot()
  })

  test('should match snapshot and check show banner when open set to true', () => {
    const { asFragment, container } = renderBanner({
      message: data.message,
      open: true,
    })
    expect(asFragment()).toMatchSnapshot()
    expect(container.firstChild).toHaveStyle(`max-height: 60px`)
  })

  test('should show read more btn when the content present update view on click', () => {
    renderBanner({
      message: data.message,
      content: data.content,
      open: true,
    })
    const moreBtn = screen.getByRole('readMore')

    expect(moreBtn).toBeInTheDocument()
    //expect(screen.getByText(data.message)).toBeInTheDocument()
    expect(moreBtn).toHaveTextContent(/read more/i)
    expect(moreBtn).not.toHaveTextContent(/read less/i)

    fireEvent.click(moreBtn)
    expect(moreBtn).not.toHaveTextContent(/read more/i)
    //expect(screen.getByText(data.content)).toBeInTheDocument()
    expect(moreBtn).toHaveTextContent(/read less/i)
  })
})
