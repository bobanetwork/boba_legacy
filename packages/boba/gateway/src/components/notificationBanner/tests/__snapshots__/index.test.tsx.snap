// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationBanner  should match snapshot and check show banner when open set to true 1`] = `
<DocumentFragment>
  <div
    class="sc-bgqQcB fPlmlQ open "
  >
    <div
      class="sc-gTRrQi hKOjqS"
      data-testid="message"
    >
      Alert message goes here
      <span
        class="sc-ewnqHT lknSpb"
        role="readMore"
      >
        Read more
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`NotificationBanner  should match snapshot without message 1`] = `
<DocumentFragment>
  <div
    class="sc-bgqQcB fPlmlQ open "
  >
    <div
      class="sc-gTRrQi hKOjqS"
      data-testid="message"
    >
      Alert message goes here
      <span
        class="sc-ewnqHT lknSpb"
        role="readMore"
      >
        Read more
      </span>
    </div>
  </div>
</DocumentFragment>
`;
